parser.hh

```diff
@@ -388,6 +404,9 @@ struct ParserPosition
 	///iterator into the input.
 	Input::iterator it;
 
+	///user-meaningful filename.
+	std::string filename;
```

Comment:
`ParserPosition` objects are only valid for the lifetime of the `Input` that they refer to.  They're also meant to be quite small (we create a lot of them!) and adding the string here makes them a lot larger - `std::string` in libc++ (64-bit) is 24 bytes, but if the filename is immutable in the input then a const reference is only 8.